src/bucket.ts

```diff
@@ -2039,15 +2040,32 @@ class Bucket extends ServiceObject {
       });
     };
 
-    this.getFiles(query)
-      .then(([files]) => {
+    (async () => {
+      try {
+        let promises = [];
         const limit = pLimit(MAX_PARALLEL_LIMIT);
-        const promises = files!.map(file => {
-          return limit(() => deleteFile(file));
-        });
-        return Promise.all(promises);
-      })
-      .then(() => callback!(errors.length > 0 ? errors : null), callback!);
+        const filesStream = this.getFilesStream(query);
+
+        for await (const curFile of filesStream) {
+          if (promises.length >= MAX_QUEUE_SIZE) {
```

Comment:
There is both `MAX_QUEUE_SIZE` and `MAX_PARALLEL_LIMIT`. Do we need both as they are trying to accomplish the same thing? Limiting the number of files being deleted at once.

Reply:
Full disclosure I'm not really sure why the original authors decided to utilize `pLimit` for this one particular function. That said, `MAX_PARALLEL_LIMIT` represents how many concurrent promises will be allowed to execute and I'm not sure why it was set to 10. We are having some discussion around if we want to expose this to be user setable. In unrelated testing we have found that setting concurrent promises really high / uncapped led to other problems, i.e. network saturation. It is probably less of a concern here as `delete` doesn't return much in the way of data. 

I implemented `MAX_QUEUE_SIZE` as a means of further reducing memory utilization. The issue you were seeing was because 1. we fetched every file into memory during `getFiles` and 2. we then created a giant array of promises 1 for each file delete. We are also discussing allowing this to be user setable.

Reply:
I understand, when the authors leave it can be hard to figure out why they wrote something, and potentially dangerous to remove without sufficient testing. It seems after more testing a proper number between 10 and 1000 could be set, or potentially as you said making it user-configurable.